spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/BaseCatalog.java

```diff
@@ -0,0 +1,196 @@
+/*
+ *  Copyright 2024 Datastrato Pvt Ltd.
+ *  This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.spark.catalog;
+
+import com.datastrato.gravitino.Catalog;
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.Namespace;
+import com.datastrato.gravitino.exceptions.NoSuchSchemaException;
+import com.datastrato.gravitino.exceptions.NonEmptySchemaException;
+import com.datastrato.gravitino.exceptions.SchemaAlreadyExistsException;
+import com.datastrato.gravitino.rel.Schema;
+import com.google.common.base.Preconditions;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import javax.ws.rs.NotSupportedException;
+import org.apache.spark.sql.catalyst.analysis.NamespaceAlreadyExistsException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.NonEmptyNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.Column;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.NamespaceChange;
+import org.apache.spark.sql.connector.catalog.SupportsNamespaces;
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.catalog.TableChange;
+import org.apache.spark.sql.connector.expressions.Transform;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+/**
+ * BaseCatalog is the base class for the specify catalog like GravitinoHiveCatalog, it provides a
+ * basic table and namespace interfaces. The advanced interfaces like view and function should be
+ * provided by the specify catalog if necessary.
+ */
+public abstract class BaseCatalog implements TableCatalog, SupportsNamespaces {
```

Comment:
Are you going to create different catalogs for different tables, like HiveCatalog, JDBC catalog, Iceberg catalog etc? 

I think for Gravitino, the table definition is unified, so it will be possible that codes are duplicated for different catalogs?

Reply:
yes, for Gravitino, the table definition is unified but the the table capacity is not. So I introduced [`GravitinoBaseTable`](https://github.com/FANNG1/gravitino/blob/table/spark-connector/src/main/java/com/datastrato/gravitino/spark/table/GravitinoBaseTable.java) to support basic schema and IO operations to reduce the code duplication.